Crypto/Number/Prime.hs

```diff
@@ -37,25 +38,54 @@ import Data.Bits
 isProbablyPrime :: Integer -> Bool
 isProbablyPrime !n
     | any (\p -> p `divides` n) (filter (< n) firstPrimes) = False
-    | primalityTestFermat 50 (n`div`2) n                   = primalityTestMillerRabin 30 n
+    | primalityTestFermat iters halfn n                    = primalityTestMillerRabin 30 n
     | otherwise                                            = False
+    where
+        halfn = n `div` 2
+        iters = fromInteger (min 50 halfn)
 
--- | generate a prime number of the required bitsize
+-- | generate a prime number of the required bitsize (i.e. in the range
+--   [2^(b-1)+2^(b-2), 2^b)).
+--
+--   May throw a CryptoError_PrimeSizeInvalid if the requested size is less
+--   than 5 bits, as the smallest prime meeting these conditions is 29.
+--   This function requires that the two highest bits are set, so that when
+--   multiplied with another prime to create a key, it is guaranteed to be of
+--   the proper size.
 generatePrime :: MonadRandom m => Int -> m Integer
 generatePrime bits = do
-    sp <- generateParams bits (Just SetTwoHighest) True
-    return $ findPrimeFrom sp
+    if bits < 5 then
+        throwCryptoError $ CryptoFailed $ CryptoError_PrimeSizeInvalid
+    else do
+        sp <- generateParams bits (Just SetTwoHighest) True
+        let prime = findPrimeFrom sp
+        if prime < 2^bits then
+            return $ prime
+        else do
+            ret <- generatePrime bits
+            return $ ret
```

Comment:
Using `numBits` from `Crypto.Number.Basic` to test the bit size would probably be more efficient (with GMP at least).

Also
```haskell
else do ret <- generatePrime bits
        return $ ret
```
can be shorten to:
```haskell
else generatePrime bits
```